backend/src/main/scala/com/softwaremill/adopttapir/starter/content/ContentModel.scala

```diff
@@ -1,11 +1,29 @@
 package com.softwaremill.adopttapir.starter.content
 
-sealed trait Node {
+import cats.syntax.all.*
+import io.circe.{Encoder, Decoder}
+import io.circe.*
+import io.circe.parser.*
+import io.circe.syntax.*
+import io.circe.generic.semiauto.*
+import sttp.tapir.Schema
+import io.circe.*
+
+sealed trait Node: // derives Codec.AsObject:
   def name: String
-}
 
-case class File(name: String, content: String) extends Node
-case class Directory(name: String, content: List[Node]) extends Node {
+object Node:
```

Comment:
I see you've moved from defining all encoders/decoders externally (as part of `Json`), to defining them as part of the ADT. This works here, since both are in the same sub-project, but I think if we do so, we should do it consistently. For example, `Schema` derivation should be included here as well (instead of being part of `Tapir`). Plus sometimes, the encoders/decoders are defined as implicit values, and sometimes using `derives`.

Reply:
It's mostly due to some limitations of `circe 0.14` for `Scala 3`. I have included `Schema` derivation in case classes but I think I need to keep the `implicit` instance of `Node` due to https://tapir.softwaremill.com/en/latest/endpoint/schemas.html#derivation-for-recursive-types-in-scala3 : when deriving `Schema` for `Node` (i.e. no `implicit` instance of `Schema[Node]`), it compiles but then seems to fall in endless loop at runtime.

Reply:
Comment added.